processor/deltatocumulativeprocessor/internal/delta/delta_test.go

```diff
@@ -0,0 +1,198 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package delta_test
+
+import (
+	"fmt"
+	"math/rand"
+	"strconv"
+	"sync"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor/internal/data"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor/internal/delta"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor/internal/streams"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor/internal/testdata/random"
+)
+
+var result any
+
+func BenchmarkAccumulator(b *testing.B) {
+	acc := delta.Numbers()
+	sum := random.Sum()
+
+	bench := func(b *testing.B, nstreams int) {
+		nsamples := b.N / nstreams
+
+		ids := make([]streams.Ident, nstreams)
+		dps := make([]data.Number, nstreams)
+		for i := 0; i < nstreams; i++ {
+			ids[i], dps[i] = sum.Stream()
+		}
+
+		b.ResetTimer()
+
+		var wg sync.WaitGroup
+		for i := 0; i < nstreams; i++ {
+			wg.Add(1)
+			go func(id streams.Ident, num data.Number) {
+				for n := 0; n < nsamples; n++ {
+					num.SetTimestamp(num.Timestamp() + 1)
+					val, err := acc.Aggregate(id, num)
+					if err != nil {
+						panic(err)
+					}
+					result = val
+				}
+				wg.Done()
+			}(ids[i], dps[i])
+		}
+
+		wg.Wait()
+	}
+
+	nstreams := []int{1, 2, 10, 100, 1000}
+	for _, n := range nstreams {
+		b.Run(strconv.Itoa(n), func(b *testing.B) {
+			bench(b, n)
+		})
+	}
+}
+
+// verify the distinction between streams and the accumulated value
+func TestAddition(t *testing.T) {
+	acc := delta.Numbers()
+	sum := random.Sum()
+
+	type Idx int
+	type Stream struct {
+		idx Idx
+		id  streams.Ident
+		dp  data.Number
+	}
+
+	streams := make([]Stream, 10)
+	for i := range streams {
+		id, dp := sum.Stream()
+		streams[i] = Stream{
+			idx: Idx(i),
+			id:  id,
+			dp:  dp,
+		}
+	}
+
+	want := make(map[Idx]int64)
+	for i := 0; i < 100; i++ {
+		stream := streams[rand.Intn(10)]
+		dp := stream.dp.Clone()
+		dp.SetTimestamp(dp.Timestamp() + pcommon.Timestamp(i))
+
+		val := int64(rand.Intn(255))
+		dp.SetIntValue(val)
+		want[stream.idx] += val
+
+		got, err := acc.Aggregate(stream.id, dp)
+		require.NoError(t, err)
+
+		require.Equal(t, want[stream.idx], got.IntValue())
+	}
+}
+
+// verify that start + last times are updated
+func TestTimes(t *testing.T) {
+	acc := delta.Numbers()
+	id, data := random.Sum().Stream()
+
+	start := pcommon.Timestamp(1234)
+	last := start
+	sum := int64(0)
+	for i := 0; i < 10; i++ {
```

Comment:
Does it need to be on a loop? Wouldn't you get the same effect with a single aggregate call?

Reply:
Unless the random aspect is relevant for the "business" code, it would just make the test more confusing to read. The test is basically the spec for the underlying code: keeping it simple will help understand what really influences the code.

Reply:
will do!